packages/fileeditor/src/fileeditorlspadapter.ts

```diff
@@ -0,0 +1,190 @@
+import { DocumentRegistry } from './../../docregistry/src/registry';
+import { CodeEditor } from '@jupyterlab/codeeditor';
+import { CodeMirrorEditor } from '@jupyterlab/codemirror';
+import { IDocumentWidget } from '@jupyterlab/docregistry';
+import {
+  IAdapterOptions,
+  IVirtualPosition,
+  VirtualDocument,
+  WidgetLSPAdapter
+} from '@jupyterlab/lsp';
+import { FileEditor } from './widget';
+
+export interface IFileEditorAdapterOptions extends IAdapterOptions {
+  /**
+   * The document registry instance.
+   */
+  docRegistry: DocumentRegistry;
+}
+
+export class FileEditorAdapter extends WidgetLSPAdapter<
+  IDocumentWidget<FileEditor>
+> {
+  constructor(
+    editorWidget: IDocumentWidget<FileEditor>,
+    options: IFileEditorAdapterOptions
+  ) {
+    const { docRegistry, ...others } = options;
+    super(editorWidget, others);
+    this.editor = editorWidget.content;
+    this._docRegistry = docRegistry;
+    this.ready = new Promise<void>((resolve, reject) => {
+      this.initOnceReady().then(resolve).catch(reject);
+    });
+
+    // Dispose the adapter when the editor is disposed.
+    editorWidget.disposed.connect(() => this.dispose());
+  }
+
+  /**
+   * The wrapped `FileEditor` widget.
+   */
+  readonly editor: FileEditor;
+
+  /**
+   * Get current path of the document.
+   */
+  get documentPath(): string {
+    return this.widget.context.path;
+  }
+
+  /**
+   * Get the mime type of the document.
+   */
+  get mimeType(): string {
+    const codeMirrorMimeType = this.editor.model.mimeType;
+    const contentsModel = this.editor.context.contentsModel;
+
+    // when MIME type is not known it defaults to 'text/plain',
+    // so if it is different we can accept it as it is
+    if (codeMirrorMimeType != 'text/plain') {
+      return codeMirrorMimeType;
+    } else if (contentsModel) {
+      // a script that does not have a MIME type known by the editor
+      // (no syntax highlight mode), can still be known by the document
+      // registry (and this is arguably easier to extend), so let's check it
+      // just in case; this is also how the "Klingon" language for testing
+      // gets registered, so we need it for tests too.
```

Comment:
```suggestion
      // registry (and this is arguably easier to extend).
```

I think there are no integration tests with Klingon in core, so let's cut it.